src/adjtrans.jl

```diff
@@ -71,13 +71,13 @@ function show(io :: IO, op :: ConjugateLinearOperator)
   show(io, op.parent)
 end
 
-function *(op :: AdjointLinearOperator, v :: AbstractVector)
-  length(v) == size(op.parent, 1) || throw(LinearOperatorException("shape mismatch"))
-  p = op.parent
+function *(op :: AdjointLinearOperator{T}, v :: AbstractVector{S}) where {T,S}
+  p = op.parent::AbstractLinearOperator{T}
```

Comment:
It seems that `AbstractLinearOperator{T}` is too generic for the compiler.
I just replace it by : 
```julia
p = op.parent::Union{LinearOperator{T}, PreallocatedLinearOperator{T}}
```
and `increase_nctprod(p)` plus `length(v) == size(p, 1)` stop allocating 16 bits at each call.

Reply:
That's too specific though. Maybe we need to parametrize AdjointLinearOperator with the type of the parent?

Reply:
I've included a new commit with the parametrized version.

Reply:
Thanks. Does it resolve the allocation issue?

Reply:
It looks like it does, but I'll wait for Alexis' tests.

Reply:
It works with my Krylov tests. :metal:  Thank you Abel! You can merge it and add a new tag.

Reply:
Before merging, would you mind checking that this doesn't recreate the hvcat issues from #97?

Reply:
N = 16
```julia
@time include("lops.jl")
  0.208034 seconds (621.17 k allocations: 31.138 MiB, 5.06% gc time)
  0.574005 seconds (1.23 M allocations: 64.762 MiB, 1.83% gc time)
5-element Array{Float64,1}:
 -10.027547809205767 
   6.387858545582767 
   4.945048065166462 
   9.668653838837873 
   3.2258245551624194
```

N = 25
```julia
@time include("lops.jl")
  0.212824 seconds (621.21 k allocations: 31.152 MiB, 4.82% gc time)
  0.561557 seconds (1.23 M allocations: 64.921 MiB, 1.83% gc time)
5-element Array{Float64,1}:
 10.628747752647548 
 -2.02582122181704  
 -4.198864112230334 
  3.0656881247622376
 10.586622803189744 
```

N = 50
```julia
@time include("lops.jl")
  0.221327 seconds (621.31 k allocations: 31.201 MiB, 4.93% gc time)
  0.592618 seconds (1.23 M allocations: 64.986 MiB, 1.84% gc time)
5-element Array{Float64,1}:
 -24.72968159374163  
 -17.87085267785233  
  21.676648685374687 
   8.68226076917717  
   2.8795559248799285
```

N = 100
```julia
@time include("lops.jl")
  0.213366 seconds (621.51 k allocations: 31.370 MiB, 4.85% gc time)
  0.586298 seconds (1.24 M allocations: 65.185 MiB, 1.76% gc time)
5-element Array{Float64,1}:
  -6.961226546652531 
  27.177113229951946 
  -3.5866280698379254
  23.01856931520228  
 -23.290794085006873 
```

Reply:
Thank you!